File: TopCoder/Aaagmnrs_TC.py
#/usr/bin/python3

class Aaagmnrs:
    def __init__(self):
        self.results = []
    def anagrams(self, ss):
        res = []
        mm = {}
        for element in ss:
            alph = {}
            sum = 0
            for char in element:
                sum += ord(char.upper())
                if ord(char.upper())-ord('A') not in alph:
                    alph[ord(char.upper())-ord('A')] = 1
                else:
                    alph[ord(char.upper())-ord('a')] = alph[ord(char.upper())-ord('A')] + 1

            if sum in mm:
                broke = False
                for alph_ in mm[sum]:
                    for char, qty in alph_.items():
                        if alph[char] is not qty:
                            broke = True
                            break
                    if broke:
                       break
                if not broke:
                    res.append(alph)
                mm[sum].append(alph)
            else:
                temp = []
                temp.append(alph)
                mm[sum] = temp
                res.append(element)
        self.results.append(res)
        return res 

File: TopCoder/Aaagmnrs_TC.cpp
#include <map>
#include <vector>
#include <string>

class Aaagmnrs{
	std::string[] anagrams(std::string ss[]){
		std::vector<std::string> res;
		map<int, std::vector<std::vector<int> > > mm;
		for(element : ss){
			std::vector<int> alph(26, 0);
			int sum = 0;
			for(int i = 0; i < element.size(); i++){
				sum += toupper(element[i]);
				alph[toupper(element[i])-'A']++;	
			}
			if(mm.count(sum)){
				bool broke = false;
				for(int i = 0; i < mm[sum].size(); i++){
					for(int k = 0; k < mm[sum][i].size(); k++){
						if(mm[sum][i][k] != alph[k]){
							broke = true;
							break;
						}	
					}
					if(broke){
						break;
					}
				}	
				if(!broke) res.push_back(element);
			}else res.push_back(element);
			mm[sum].push_back(alph);
		}	
		std::string rs[res.size()];
		for(int i = 0; i < res.size(); i++) rs[i] = res[i];
		return rs;
	}
};
